### DISCUSSION TITLE
Use the packagist metadata-changes-url API for use the cache more efficiently.

### DISCUSSION THREAD

**Discussion by vtsykun** [composer/composer#11330] [General]
I noticed that the packagist provide metadata-changes-url API endpoint `https://packagist.org/metadata/changes.json` 
Example response:
```json
{
  "actions": [
    {
      "type": "update",
      "package": "kadirov/api-starter-kit~dev",
      "time": 1676476155
    },
    {
      "type": "update",
      "package": "automattic/vip-go-mu-plugins~dev",
      "time": 1676476160
    },
    {
      "type": "update",
      "package": "adyen/adyen-shopware6~dev",
      "time": 1676476161
    },
    {
      "type": "update",
      "package": "defstudio/laravel-components",
      "time": 1676476161
    },
    {
      "type": "update",
      "package": "defstudio/laravel-components~dev",
      "time": 1676476161
    }
  ],
  "timestamp": 16764761612254
}
```

As far as I understand the original purpose of this API use it for synchronization mirrors. Can composer use this API to request only packages that have been updated since the command was last run?

**Comment by stof** [comment]
> Can composer use this API to request only packages that have been updated since the command was last run?

Which command are you talking about ?

**Reply by vtsykun** [reply]
For example, composer `update` 
composer requests all packages, can we avoid unnecessary 304 requests with use metadata-changes-url api?
 
![Selection_1286](https://user-images.githubusercontent.com/21358010/219092508-ab83ced6-a8c6-4a32-9cf9-4a0fb2bb7b4d.png)

**Reply by stof** [reply]
I don't think you can use metadata-changes-url.
Your cache does not contain a full mirror of the packagist metadata (very far from that) and each cache file might have been requested at a different time and so have a different Last-Modified-Date.
And the metadata-changes endpoint only supports querying about changes that are recent enough. If your mirror is too outdated, it will send you a single entry in the metadata changes asking to run a full mirroring again.

Trying to avoid the packagist.org queries by abusing the metadata-changes endpoint will require lots of complexity and I don't think it will be faster.

Those 304 responses correspond to static files, served directly by nginx/Apache in each of the repo.packagist.org locations (that domain corresponds to multiple servers in different locations). The metadata-changes endpoint is a dynamic endpoint that is served by the Symfony application powering packagist.org, which requires reaching the main datacenter and is slower to respond (and for any file where the cache is not fresh, you would still have to query that file).
Thus, when you have ext-curl on your system, those HTTP requests are done in parallel, while any implementation based on metadata-changes-url would have to wait for the result of those changes before starting to do any "optimized" fetching based on the state of the composer cache.
